generate_workspace/src/main/java/com/google/devtools/build/workspace/maven/GraphSerializer.java

```diff
@@ -0,0 +1,82 @@
+package com.google.devtools.build.workspace.maven;
+
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.google.common.graph.GraphBuilder;
+import com.google.common.graph.MutableGraph;
+import org.eclipse.aether.graph.DependencyNode;
+import org.eclipse.aether.graph.DependencyVisitor;
+
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Serializes a resolved dependency graph into a set of Bazel Build Rules.
+ */
+public class GraphSerializer {
+
+  public static Set<MavenJarRule> generateBuildRules(DependencyNode root) {
+    MutableGraph<MavenJarRule> graph = GraphBuilder.directed().allowsSelfLoops(false).build();
+    DependencyVisitor visitor = new AetherGraphTraverser(graph);
+    root.getChildren().forEach(c -> c.accept(visitor));
+    return graph.nodes();
+  }
+}
+
+/**
+ * Performs a depth first traversal of aether's dependency node graph, and
+ * constructs a guava graph representation of the aether graph, with each
+ * node being a Rule object.
+ */
+class AetherGraphTraverser implements DependencyVisitor {
+  private final Set<DependencyNode> visited;
+  private final MutableGraph<MavenJarRule> ruleGraph;
+  private final Map<DependencyNode, MavenJarRule> nodeRuleMap;
+
+  AetherGraphTraverser(MutableGraph<MavenJarRule> ruleGraph) {
+    this.visited = Sets.newHashSet();
+    this.ruleGraph = ruleGraph;
+    this.nodeRuleMap = Maps.newHashMap();
+  }
+
+  @Override
+  public boolean visitEnter(DependencyNode dependencyNode) {
+    if (dependencyNode == null || dependencyNode.getDependency().isOptional()) {
+      return false;
+    }
+    boolean isFirstVisit = visited.add(dependencyNode);
+    if (!isFirstVisit) {
```

Comment:
In my experience, Maven/Aether does not define equals() methods, so probably `new DependencyNode() != new DependencyNode()`.